File: LeetCode/Go/lengthOfLongestSubstring/lengthOfLongestSubstring.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	left := 0
	maxLength := 0

	// character -> last index
	charMap := make(map[byte]int)

	for right := 0; right < len(s); right++ {

		if index, exists := charMap[s[right]]; exists && index >= left {
			left = index + 1
		}

		charMap[s[right]] = right

		if right-left+1 > maxLength {
			maxLength = right - left + 1
		}
	}

	return maxLength
}

func lengthOfLongestSubstringUsingset(s string) int {
	set := make(map[byte]bool)
	left := 0
	maxLength := 0

	for right := 0; right < len(s); right++ {

		for set[s[right]] {
			delete(set, s[left])
			left++
		}

		set[s[right]] = true

		if right-left+1 > maxLength {
			maxLength = right - left + 1
		}
	}

	return maxLength
}

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb")) // 3
	fmt.Println(lengthOfLongestSubstring("bbbbb"))    // 1
	fmt.Println(lengthOfLongestSubstring("pwwkew"))   // 3
	fmt.Println(lengthOfLongestSubstring(""))         // 0
	fmt.Println(lengthOfLongestSubstring("abcabcbb")) // 3
}

File: LeetCode/Javascript/lengthOfLongestSubstring.js
//Longest Substring Without Repeating Characters
//Given a string s, find the length of the longest substring without duplicate characters.

/*
Example 1:

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3. Note that "bca" and "cab" are also correct answers.
Example 2:

Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
Example 3:

Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
*/


function lengthOfLongestSubstring(s) {
    let map = new Map();
    let left = 0;
    let maxLength = 0;

    for (let right = 0; right < s.length; right++) {
        // If character already exists in current window
        if (map.has(s[right]) && map.get(s[right]) >= left) {
            left = map.get(s[right]) + 1;
        }
        map.set(s[right], right);
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
}

// Examples
//console.log(lengthOfLongestSubstring("abcabcbb")); // 3
// console.log(lengthOfLongestSubstring("bbbbb"));    // 1
 console.log(lengthOfLongestSubstring("pwwkew"));   // 3
// console.log(lengthOfLongestSubstring(""));         // 0



function lengthOfLongestSubstringUsingSet(s) {
    let set = new Set();
    let left = 0;
    let maxLength = 0;

    for (let right = 0; right < s.length; right++) {

        while (set.has(s[right])) {
            set.delete(s[left]);
            left++;
        }

        set.add(s[right]);
        console.log(set)
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
}

console.log(lengthOfLongestSubstringUsingSet("abcabcbb"));
 